5-spring-security-integration

vanyaserdyuk opened the pull request.

vanyaserdyuk committed:
Spring Security Integration #5

iv21dk commented:
`Revert "stopwatch timer"` - for what?

iv21dk commented on .idea/compiler.xml:
```diff
@@ -2,6 +2,7 @@
 <project version="4">
```
Please remove this file from git

iv21dk commented on src/main/java/ru/testtask/controller/MainController.java:
```diff
@@ -81,4 +91,21 @@
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PutMapping("/user/{id}")
+    public ResponseEntity<User> updateUser(@PathVariable("id") String id, @RequestBody User clientUser) {
```
Please create new controller `UserController` and move this method

iv21dk replied:
`User clientUser` - please create and use DTO

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,47 @@
+        User user = userRepo.findByUsername(username);
+
+        if (user == null){
+            throw new UsernameNotFoundException("User does not exists");
```
Please add `username` to message using String.format()

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public String getCurrentUserId(){
+        org.springframework.security.core.userdetails.User user = (org.springframework.security.core.userdetails.User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
```
The method can be used from thread that doesn't know about current context. We need to check that context is available. 
Please call following from controller:
```
new Thread(() -> {
            org.springframework.security.core.userdetails.User user = (org.springframework.security.core.userdetails.User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        }).start();
```

iv21dk commented on src/main/java/ru/testtask/controller/MainController.java:
```diff
@@ -55,6 +64,7 @@
         return new ResponseEntity<>(projects, HttpStatus.OK);
     }
 
+    @PreAuthorize("projectService.findProjectById(#id).getOwnerId() == userService.getCurrentUserId() or hasAuthority('MODERATOR')")
```
You can create new method ProjectService#isOwner(String userId) and use intead

iv21dk reviewed (changes requested):
I don't found api to login.

iv21dk commented on src/main/java/ru/testtask/model/User.java:
```diff
@@ -0,0 +1,54 @@
+public class User implements UserDetails {
+    @Id
+    private String id;
+    private String username;
```
`username` should be unique filed

iv21dk replied:
Please add new test to `ProjectServiceTest` to check that impossible to create two projects with same names

iv21dk commented on src/main/java/ru/testtask/dto/ClientUserDTO.java:
```diff
@@ -0,0 +1,16 @@
+
+@Data
+@NoArgsConstructor
+public class ClientUserDTO {
```
`UserDTO` is enough to understand the purpose of the class

iv21dk commented on src/main/java/ru/testtask/exception/WrongMethodUseException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.testtask.exception;
+
+public class WrongMethodUseException extends Exception {
```
extends `RuntimeException` is better. We don't need to catch this exception.

iv21dk commented on src/main/java/ru/testtask/service/ProjectService.java:
```diff
@@ -58,5 +64,23 @@ public void updateProject(Project project) {
         return projectRepo.findAll();
     }
 
+    public boolean isOwner(String id){
```
I suggested the wrong method name. Let's rename it to `isCurrentUserOwnerOf(String projectId)`

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -17,24 +19,30 @@
     @Autowired
     private UserRepo userRepo;
 
+    @SneakyThrows
```
I don't see reason to add this annotation. We need to disscuss it

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -17,24 +19,30 @@
-        return getUserByUsername(user.getUsername()).getId();
+    public String getCurrentUserId() throws WrongMethodUseException{
+        try {
+            org.springframework.security.core.userdetails.User user = (org.springframework.security.core.userdetails.User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
```
It is better to check `SecurityContextHolder.getContext().getAuthentication()` value. If value is null you can throw exception with message 
`Method can't be used from this thread. `

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,55 @@
+            return getUserByUsername(user.getUsername()).getId();
+        }
+        catch (Exception e){
+            throw new WrongMethodUseException("Method can't be use");
```
Please add `e` as additiolal parameter to constructor of `WrongMethodUseException`

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,54 @@
+        if (SecurityContextHolder.getContext().getAuthentication() == null)
+            throw new WrongMethodUseException("Method can't be used from this thread.");
+
+        else return null;
```
I expect method will return user id

iv21dk commented on src/main/java/ru/testtask/service/ProjectService.java:
```diff
@@ -54,5 +59,18 @@ public void updateProject(Project project) {
+    public boolean isCurrentUserOwnerOf(String id){
+        Project project;
+
+            Optional<Project> optionalProject = findProjectById(id);
```
Please format this code

iv21dk commented on src/main/java/ru/testtask/service/ProjectService.java:
```diff
@@ -54,5 +59,18 @@ public void updateProject(Project project) {
+            }
+            else return false;
+
+           return userService.getCurrentUserId().equals(project.getOwnerId());
```
`Objects.equals()` is better because it more safe in case if `userService.getCurrentUserId()` will return null

iv21dk commented on src/main/java/ru/testtask/converter/DTOConverterConfig.java:
```diff
@@ -0,0 +1,25 @@
+
+import static org.modelmapper.config.Configuration.AccessLevel.PRIVATE;
+
+@Component
```
Should be @Configuration

vanyaserdyuk committed:
Fixed merging issues

iv21dk commented on src/main/java/ru/testtask/controller/MainController.java:
```diff
@@ -1,26 +1,36 @@
     private ProjectService projectService;
 
+    @Autowired
+    private UserService userService;
```
Please remove unused field

iv21dk commented on src/main/java/ru/testtask/controller/MainController.java:
```diff
@@ -45,15 +55,22 @@
 
     @PostMapping()
-    public ResponseEntity<ProjectDTO> postProject(@RequestBody CreateProjectDTO createProjectDTO) {
+    public ResponseEntity<?> postProject(@RequestBody CreateProjectDTO createProjectDTO) {
```
Old version is better

iv21dk commented on src/main/java/ru/testtask/controller/MainController.java:
```diff
@@ -70,21 +87,22 @@
 
         if (projectService.findProjectById(id).isEmpty()) {
-            return new ResponseEntity<>(String.format("Project with ID %s does not found", id), HttpStatus.NOT_FOUND);
+            return new ResponseEntity<>("This project does not exist!", HttpStatus.NOT_FOUND);
```
Old version is better

iv21dk commented on src/main/java/ru/testtask/model/Project.java:
```diff
@@ -9,8 +12,6 @@
 
 @Document(collection = "projects")
 @NoArgsConstructor
-@Builder
```
I think it is better to keep these annotations

iv21dk commented on src/main/java/ru/testtask/service/ProjectService.java:
```diff
@@ -30,13 +38,20 @@ public Project findProjectByName(String name){
-            project.addGeometry(Geometry.builder().name("geometry" + i).id(UUID.randomUUID().toString()).build());
-            project.addAttribute(Attribute.builder().name("geometry" + i).id(UUID.randomUUID().toString()).build());
+    public Project createProject(Project project) throws NameAlreadyExistsException {
+        if (findProjectByName(project.getName()).getId() != null){
```
I think this code will work only in case if you try to create new project with conflict name (if project with same name already exists)
`findProjectByName(project.getName())` will return `null` in any other cases

iv21dk commented on src/main/java/ru/testtask/service/ProjectService.java:
```diff
@@ -30,13 +38,20 @@ public Project findProjectByName(String name){
+                project.addAttribute(Attribute.builder().name("geometry" + i).id(UUID.randomUUID().toString()).build());
+            }
+            project.setOwnerId(userService.getCurrentUserId());
+            projectRepo.insert(project);
```
`return projectRepo.insert(project);` is better because you will not make an additional request to db

iv21dk commented on src/main/java/ru/testtask/service/ProjectService.java:
```diff
@@ -30,13 +38,20 @@ public Project findProjectByName(String name){
+            throw new NameAlreadyExistsException("Project with the same name already exists!");
+        }
+
+        else {
```
Please use standard formatting

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException{
+            User user = userRepo.findByUsername(username);
```
Please format this code

iv21dk commented on src/test/java/ProjectServiceTest.java:
```diff
@@ -47,4 +47,11 @@ public void createTest(){
+    @Test
+    public void checkSimilarNamesCreationTest(){
+        Project resultProject = projectService.createProject(testProject);
+        Project resultProject2 = projectService.createProject(testProject);
```
Second calling of projectService.createProject(testProject) will failed and exception will throw. It is correct. Your test should check that particular exception is really throw.

iv21dk commented:
```
Following users should be created automatically if they are not exist:
...
```

The issue requirement was ignored

vanyaserdyuk committed:
not final

vanyaserdyuk committed:
I have a question about a comment
"I think this code will work only in case if you try to create new project with conflict name (if project with same name already exists)
 findProjectByName(project.getName()) will return null in any other cases"

 This code is needed for this case, to forbid users from creating projects with same names. I dont understand what should I do with this

iv21dk commented on src/main/java/ru/testtask/Application.java:
```diff
@@ -3,19 +3,31 @@
+
+    @PostConstruct
+    public void createUsers(){
+        userService.createDefaultUsers();
```
It is better to move this method to `UserService` or just add `@PostConstruct` to `UserService#createDefaultUsers()`

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public void createDefaultUsers(){
+        User user = User.builder().username("user").password("user").roles(Collections.singleton(Role.USER)).build();
```
`user` instance can be created only in case if it isn't exist (before `userRepo.insert(user);`)

iv21dk commented on src/main/java/ru/testtask/service/UserService.java:
```diff
@@ -0,0 +1,77 @@
+            userRepo.insert(user);
+        }
+
+        User admin = User.builder().username("admin").password("admin").roles(Collections.singleton(Role.ADMIN)).build();
```
I see duplicated code. It is better to extract this logic to additional method

vanyaserdyuk committed:
Changed the logic of default users creation

iv21dk commented on pom.xml:
```diff
@@ -104,6 +114,18 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
I don't see reason to add these dependencies

vanyaserdyuk committed:
Some changes

vanyaserdyuk committed:
Some changes
